common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStream.scala

```diff
@@ -132,14 +133,53 @@ object PipelineStorageStream extends Logging {
     msgSender: MessageSender[MsgDestination],
     indexer: Indexer[T])(implicit
                          ec: ExecutionContext,
-                         indexable: Indexable[T]) =
-    Flow[(Message, Option[T])]
-      .collect { case (msg, Some(document)) => Bundle(msg, document) }
+                         indexable: Indexable[T]) = {
+    val maxSubStreams = Integer.MAX_VALUE
+    Flow[(Message, List[T])]
+      .collect {
+        case (msg, items@_ :: _) =>
+          items.map(item =>
+            Bundle[T](message = msg, item = item, numberOfItems = items.size))
+      }
+      .mapConcat[Bundle[T]](identity)
       .via(batchIndexFlow(config, indexer))
-      .mapAsyncUnordered(config.parallelism) {
-        case Bundle(msg, id) =>
-          Future.fromTry(msgSender.send(id).map(_ => msg))
+      .via(groupByMessage(maxSubStreams, 5 minutes).mergeSubstreams)
+      .mapConcat(identity)
+      .mapAsyncUnordered(config.parallelism) { bundle =>
+        for {
+          _ <- Future.fromTry(msgSender.send(indexable.id(bundle.item)))
+        } yield bundle
+      }
+      .via(groupByMessage[T](maxSubStreams, 5 minutes)
+        .collect {
+          case head :: _ => head.message
+        }
+        .mergeSubstreams)
+  }
+
+  // Splits the flow into a substream for each messageId.
+  // Each substream emits one message with the complete list of bundles for the same messageId
+  // or no message if it didn't receive the correct number of bundles
+  def groupByMessage[T](maxSubStreams: Int, t: FiniteDuration) =
+    Flow[Bundle[T]]
+      .groupBy(maxSubstreams = maxSubStreams, f = _.message.messageId(), allowClosedSubstreamRecreation = true)
+      .scan(Nil: List[Bundle[T]]) {
+        case (bundleList, b) => b :: bundleList
+      }
+      .filter { list =>
+        list.nonEmpty && list
+          .map(_.item)
+          .distinct
+          .size == list.head.numberOfItems
       }
+    // Close the substream with take if it succeeds
+    // or after timeout if it fails
```

Comment:
How does this resolve the problem that we saw before?

Reply:
The problem was that there's a maximum number of substreams that you can have. I was setting it to Integer.MAX_VALUE, but because I was never closing them, eventually it would be unable to open substreams